woocommerce/woocommerce-ios
# Feature/add release tooling

This PR adds some Fastlane lanes that help the release of the app.
In order to share the same code, all the actions in the WordPress iOS repository have been extracted and added to the release-toolkit.

# To Test
Testing is not trivial and should be done paying great attention, especially for the code_freeze lane because the test is done on the main repository.

In order to make the steps clearer, I'm referring to defined release numbers, instead of saying _this version_ or _next version_. Note that these numbers works as of today, when 1.4 is the version in beta testing. If you test this PR later, you have to change the number consequently. 

### code_freeze lane
This lanes executes the steps to code freeze a new version of the app.
Please, be super careful with this and make sure to revert all the changes that the script does.

1. Tag a fake version 1.4 in your local repository: `git checkout develop`  &&  `git tag 1.4`.  You don't need push it to origin.
2. Checkout this branch and run `bundle exec fastlane code_freeze`.
3. Verify that a new 1.5 branch has been created.
4. Verify that the version has been updated in XCConfig, Deliverfile and download_metadata.
5. Verify that the English version of Localizable.strings has been updated with the latest strings.
6. Verify that the 1.5 milestone has marked with the snowflake icon.
7. Verify that the branch protection has been set for the `release/1.5` branch.
8. Verify that the change log file has been created in your home.

Clean up
1. Remove the branch protection for the `release/1.5` branch on GitHub.
2. Delete the `release/1.5` branch on GitHub.
3. Remove the snowflake icon from the 1.5 milestone.
4. Remove the 1.5 tag from your local repository.
5. Remove the `release/1.5` branch from your local repository.

### new_beta_release lane
This lane updates the version of the app to prepare for a new beta release.
The simpler way to test this lane is to use a forked repository.
1. Fork the repository. For the rest of the steps, use the forked repo.
2. Checkout the `release/1.4` branch and merge this branch into it.
3. Run `bundle exec fastlane new_beta_release`.
4. Verify that the proposed version is right and confirm.
5. Verify that the xcconfig file has been updated correctly and that the new tag has been added.
6. Verify that all the files have been committed.

### finalize_release lane
This lane bumps the version to the final one, downloads the translations and clean the temp tags.
The simpler way to test this lane is to continue to use the forked repository from the step above.
1. Checkout the `release/1.4` branch.
2. Run `bundle exec fastlane finalize_release`.
3. Verify that the app translations and the metadata translations have been downloaded.
4. Verify that the xcconfig file has been updated correctly and that the new tag has been added.
5. Verify that all the files have been committed.
6. Verify that the temporary tags for the beta versions have been deleted.

### build_and_upload_release lane
This lane builds the current branch and uploads the binary to TestFlight and the related dSYM to Crashlytics. 
Since we can't delete test versions from TestFlight, the test can be performed rebuilding the current release/1.4 branch (the original one! not the forked one that has been configured with a new version number!) and verifying that the upload fails because a binary with that version already exists. 
It confirms that the build, signing and login into TestFlight work. The actual upload is handled by the Fastlane itself.
1. Switch to the original repository.
2. Checkout the `release/1.4` branch.
3. Run `bundle exec fastlane build_and_upload_release`.
4. Verify that the binary is built.
5. When prompted, login into TestFlight.
6. Verify that the upload is started and that it fails because of the version validation.
